### ISSUE TITLE
C# interfaces implemented from outside the wrapped library

### ISSUE THREAD

**Issue by loicmorvan** [swig/swig#1265]
Let's say I have this *C++* code:
``` c++
namespace MyLibrary {
    class INamed { public: virtual const char* GetName() = 0; };
    class Service { public: void DisplayName(INamed* named) { cout << named->GetName() << endl; } };
}
```
The purpose here is to let users of the library implement the interface `INamed`.

Now I use *swig* to generate a *C#* wrapper by including these lines in the interface file:
```
%include <swiginterface.i>
%interface_custom("NamedProxy", "INamed", MyLibrary::INamed);
```
It generates the classes and interfaces that I want (attributes are intentionally hidden):
``` csharp
public class Service { 
    public void DisplayName(INamed named) { ... }
}

public interface INamed {
    global::System.Runtime.InteropServices.HandleRef GetInterfaceCPtr();
    string GetName();
}

public class NamedProxy: INamed { ... }
```

Then I implement the `INamed` interface and use it:
``` csharp
public class MyNamedObject: INamed {
    public string GetName() => "MyNamedObject";
    HandleRef GetInterfaceCPtr() => ???;
}

public static class Program {
    public static void Main() {
        var service = new Service();
        service.DisplayName(new MyNamedObject());
    }
}
```

This program crashes with an `AccessViolationException` or `NullReferenceException` depending on the implementation of `GetInterfaceCPtr()`.

My questions are:
- What is the purpose of the `NamedProxy` class?
- What is the purpose of the `HandleRef GetInterfaceCPtr()` method? How is it supposed to be implemented?
- Does *swig* support executing a code implemented in *C#* from the *C/C++* library?

**Comment by vadz** [comment]
> Does swig support executing a code implemented in C# from the C/C++ library?

Yes, you should look at the "directors" in the documentation/examples.

**Comment by loicmorvan** [comment]
I was indeed able to implement a method in *C#* through a pure virtual method from an abstract class and it works well!

However, I want to do that with interfaces and not abstract classes (for multiple implementation purposes). And still, the generated interface has that annoying `GetInterfaceCPtr()` method to implement and a "proxy" is generated.

**Comment by ojwb** [comment]
We seem to have one C# testcase which exercises `%interface_custom` (` Examples/test-suite/csharp/multiple_inheritance_overload_runme.cs `) but it doesn't actually test implementing the created C# interface AFAICS.

@wsfulton You implemented this originally so you're probably best placed to comment on how this is meant to work.
